pkg/tbtc/registry.go

```diff
@@ -0,0 +1,195 @@
+	// wallet persistence.
+	walletStorage *walletStorage
+}
+
+// newWalletRegistry creates a new instance of the walletRegistry.
+func newWalletRegistry(persistence persistence.Handle) *walletRegistry {
+	walletStorage := newWalletStorage(persistence)
+
+	// Pre-populate the wallet cache using the wallet storage.
+	walletCache := walletStorage.loadSigners()
+	if len(walletCache) > 0 {
+		for walletStorageKey, signers := range walletCache {
+			logger.Infof(
+				"wallet signing group [0x%v] loaded from storage "+
+					"with [%v] members",
+				walletStorageKey,
+				len(signers),
+			)
+		}
+	} else {
+		logger.Infof("no wallet signing groups found in the storage")
+	}
+
+	return &walletRegistry{
+		walletCache:   walletCache,
+		walletStorage: walletStorage,
+	}
+}
+
+// registerSigner registers the given signer using in the walletRegistry.
+func (wr *walletRegistry) registerSigner(signer *signer) error {
+	wr.mutex.Lock()
+	defer wr.mutex.Unlock()
+
+	err := wr.walletStorage.saveSigner(signer)
+	if err != nil {
+		return fmt.Errorf("cannot save signer in the storage: [%w]", err)
+	}
+
+	walletStorageKey := getWalletStorageKey(signer.wallet.publicKey)
+
+	wr.walletCache[walletStorageKey] = append(
+		wr.walletCache[walletStorageKey],
+		signer,
+	)
+
+	return nil
+}
+
+// getSigners gets all signers for the given wallet held by the walletRegistry.
+func (wr *walletRegistry) getSigners(
+	walletPublicKey *ecdsa.PublicKey,
+) []*signer {
+	wr.mutex.Lock()
+	defer wr.mutex.Unlock()
+
+	return wr.walletCache[getWalletStorageKey(walletPublicKey)]
+}
+
+// walletStorage is the component that persists data of the wallets managed
+// by the given node using the underlying persistence layer. It should be
+// used directly only by the walletRegistry.
+type walletStorage struct {
+	// persistence is the handle to the underlying persistence layer.
+	persistence persistence.Handle
+}
+
+// newWalletStorage creates a new instance of the walletStorage.
+func newWalletStorage(persistence persistence.Handle) *walletStorage {
+	return &walletStorage{persistence}
+}
+
+// saveSigner saves the given signer using the underlying persistence layer
+// of the walletStorage.
+func (ws *walletStorage) saveSigner(signer *signer) error {
+	signerBytes, err := signer.Marshal()
+	if err != nil {
+		return fmt.Errorf("could not marshal signer: [%w]", err)
+	}
+
+	err = ws.persistence.Save(
+		signerBytes,
+		getWalletStorageKey(signer.wallet.publicKey),
+		fmt.Sprintf("/membership_%v", signer.signingGroupMemberIndex),
+	)
+	if err != nil {
+		return fmt.Errorf(
+			"could not save membership using the "+
+				"underlying persistence layer: [%w]",
+			err,
+		)
+	}
+
+	return nil
+}
+
+// loadSigners loads all signers stored using the underlying persistence layer.
+func (ws *walletStorage) loadSigners() map[string][]*signer {
+	signersByWallet := make(map[string][]*signer)
+
+	descriptorsChan, errorsChan := ws.persistence.ReadAll()
+
+	var wg sync.WaitGroup
+	wg.Add(2)
```

Comment:
Should we explain the logic behind two goroutines and the wait group?
```suggestion
	// Two goroutines read from descriptors and errors channels and either
	// adds signer to the cache or outputs an error to stderr.
	// The reason for using two goroutines at the same time - one for
	// descriptors and one for errors - is that channels do not have to be
	// buffered and we do not know in what order information is written to
	// channels.
	var wg sync.WaitGroup
	wg.Add(2)
```

Reply:
See 696d8753